### ISSUE TITLE
Questions about the training loss

### ISSUE THREAD

**Issue by judiebig** [YangZhengyi98/DreamRec#1]
Hi, congratulations on the acceptance of your work by NeurIPS 2023!

I have a question regarding the training objective: your work directly uses the reconstruction loss as the training objective. However, as mentioned in [1, 2], directly use MSE loss with randomly initialized token representations may lead to representation collapse.

I fully agree with the points raised in [1, 2]. In my understanding, unless you use a pre-trained embedding dictionary and fix them (stop gradient), relying solely on the reconstruction loss can not yield meaningful results -- in the original DDPMs scenario, input features (from images, audios, etc) are fixed, meaningful, and distinguishable from each other [1]. Upon reviewing your code, I noticed that you initialize item embeddings using "nn.Embedding," introducing a potential risk of a trivial solution where all item embeddings collapse to the same value. In such a scenario, with identical item embeddings, the model can easily predict the next item using previous embeddings, simultaneously reducing the loss to 0 during training. However, in the inference stage, selecting the true item becomes challenging as K-nearest items become meaningless.

From my perspective, the existing diffusion recommendation models, which enhance the training objective by incorporating a classification loss (can be used independently or in conjunction with a KL loss), seem more reasonable. 

Do I misunderstand some details? I look forward to your reply!

[1] Conditional Denoising Diffusion for Sequential Recommendation
[2] Difformer: Empowering Diffusion Model on Embedding Space for Text Generation

**Comment by YangZhengyi98** [comment]
Thank you for your valuable comments! I understand your concern that there might exist embedding collapse issue when directly using the objective function of DDPM. Here I would like to share my understanding of it:

> Inputs of DDPM (i.e., images) are fixed, meaningful and distinguishable.
 
In ID-based sequential recommendation, the inputs can be viewed as fixed one-hot vectors representing item IDs, and the 'nn.Embedding' can be viewed as a linear projection of the one-hot vector to acquire the corresponding embedding.  From this point, the inputs are also fixed and distinguishable.

> The potential risk that all item embeddings might collapse to the same value. 

From the analysis above, the collapse issue is the collapse of the linear projection layer of the recommender model. Actually, the model collapse issue is quite common in previous generative models such as GANs and VAEs, even in image generation task. The reason can be the iteration of two-stage training (min-max or encoder-decoder). In contrast, the training of diffusion models involves denoising for one step from the noised sample, which suffers less from the collapse issue since plenty of Gaussian noise is randomly sampled and utilized in the training process.

> Enhancing the training objective in DDPM with other losses.

We do believe that it is promising to design auxiliary tasks such as contrastive learning, as discussed in the conclusion section. Meanwhile, the original training objective ('MSE') in DDPM should be carefully preserved to guarantee the theoretical soundness of diffusion models. Besides, the KL-divergence term in DDPM cannot be easily compromised as inner-product of embeddings and cross-entropy without more strict derivation.

Hope this response is helpful! Please let me know if you have further comments or suggestions.

**Comment by judiebig** [comment]
Thanks for your timely response!

I'm afraid I didn't express my concerns clearly enough. I'd like to suggest you read the references I mentioned above (especially ref [1], page 4, Limitations), which illustrate the details of why they didn't use reconstruction loss only.

**Comment by YangZhengyi98** [comment]
It seems to be an open question whether the cross-entropy loss should be a must or not. Recent work (such as DIFFUSEQ [3]) does not include the reconstruction loss in the embedding-based text generation. Therefore, researchers might have not reached a clear conclusion about this question.

[3] Gong S, Li M, Feng J, et al. Diffuseq: Sequence to sequence text generation with diffusion models[J]. In ICLR 2023.

**Comment by judiebig** [comment]
Sorry for the confusion above. The reconstruction loss that I say (also mentioned in [1,2]), is the simplified KL loss term in original DDPMs -- a MSE loss to reconstruct the original input embedding. So, my question is not whether the reconstruction loss should be a must or not, but whether the reconstruction loss could be used alone without CE loss. According to my understanding and all existing DDPM-based recommendation methods, a cross-entropy (CE) loss should be incorporated to regularize the embedding learning process (unless item embeddings are pre-trained and fixed). 

I also read the paper you recommend ([3]) and find that DIFFUSEQ has a rounding loss (Eq. 1, 2). After I check their code, I find the code in diffuseq/gaussian_diffusion.py, line 632-636, has a decoder_nll loss using nn.CrossEntropy. Since DIFFUSEQ is inspired by [4], I read [4] and find that it also has a CE loss for projecting embedding into the text (Eq. 2).

It seems that all previous works (in discrete spaces, such as recommendation and NLP) require CE loss. 

My concern is that, in this work, you initialize item embeddings randomly, but only use MSE loss to optimize your model. Since there are many potential but meaningless solutions that can satisfy your training objectives, how do you guarantee that your model can learn a meaningful embedding space?

Please note that I am not very familiar with NLP, and I just quickly read the paper and check the code, so I'm wondering if I'm missing something important. Looking forward to your reply :)

[4] Diffusion-LM Improves Controllable Text Generation, NeurIPS, 2022.

**Comment by YangZhengyi98** [comment]
Hi, sorry about the ambiguity in the above response, the 'reconstruction loss' should be 'cross-entropy loss'.

According to  Equation (2) of [3], no cross-entropy loss is used, but an L-2 regularization of embedding（derivated in Equation (17) of [3] in Appendix A).  The rounding operation is only conducted at the sampling phase in [3] (detailed in the paragraph after Equation (3)).

I am also not so familiar with NLP, so I would rather view this as an open question without an absolute conclusion.

I have also tried the L-2 regularization in our experiment but observed no noticeable improvement, therefore we only use the reconstruction loss. 

The learned embeddings are well scattered as shown in the TSNE visualization in the paper, which suggests that the collapse issue may not be as serious as you might think before.

We also show that the generated oracle embeddings and the learned item embeddings can be interpreted and distinguished by Large Language Models on Movielens and Steam datasets (https://arxiv.org/abs/2310.20487), which can also verify that the learned embeddings of DreamRec are not collapsed but distinguishable.

**Comment by judiebig** [comment]
The term in Eq. (2), first line ([3]) is $- \log p_\theta\left(\mathbf{w}^{x \oplus y} \mid \mathbf{z}_0\right)$, which is represented as $\mathcal{R}\left(\left\|\mathbf{z}_0\right\|^2\right)$ in the second line.  Since they use $\text{EMB}(\cdot)$ to represent embedding, I guess $\mathbf{w}^{x \oplus y}$ is not the word embedding, but the word (or its index).

According to the code in diffuseq/gaussian_diffusion.py, line 632-636, they use:

> terms["loss"] = terms["mse"] + decoder_nll + tT_loss

and the implementation of "decoder_nll" contains (line 554):

> loss_fct = th.nn.CrossEntropyLoss(reduction='none')

I also checked the code of "get_logits" in "transformer_model.py" (line 109-124), and it seems that DIFFUSEQ indeed projects the embeddings to the word index.

Issues #25 and #27 in [3]'s repo are related with this nll_loss. From the authors' response (#25: you can regard the "decoder_nll" as a regularization term of the embedding vectors), I guess Eq. (2) use $\mathcal{R}\left(\left\|\mathbf{z}_0\right\|^2\right)$ to represent "decoder_nll" loss. This loss ensures that word embeddings should be different and distinguishable, so that model could learn a meaningful embedding space. 

Can you check their code again? If I misunderstand the code, I would really appreciate your correction. Thanks :)

**Comment by YangZhengyi98** [comment]
Hi,  let us analyze the code of [3] more carefully.

The 'terms["mse"]' denotes the reconstruction loss. The decoder_nll is the regularization loss, which is based on 'x_tart', the word embedding  BEFORE the diffusion steps, not the sampled embeddings from the Gaussian noise such as [4] . "get_logits" in "transformer_model.py" (line 109-124) is the caculation of L-2 distances, which corresponds to the $\mathcal{R}\left(\left\|\mathbf{z}_0\right\|^2\right)$ term of Equation 2, which keeps the learned word embedddings close to the initialized ones.

We also tried the L-2 regularization and observed no noticeable improvement, as I have mentioned before. And the TSNE visualization suggests the learned embeddings are not collapsed.

Moreover, I think the  ’terms["nll"]‘ term (line 633, diffuseq/gaussian_diffusion.py of [3]) is the cross-entropy loss you keep mentioning in the above questions, which is the rounding loss in [4]. But this term is NOT used in [3]. I hope you notice this difference.

Besides, Neither [3] nor [4] mention the collapse issue, so I would not reach any easy conclusion nut view this as an open question.

I sincerely recommend more careful investigation, and interpreting paper and code of other publications line by line here is really beyond my responsibility.

**Comment by judiebig** [comment]
Thanks a lot :)
